src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.cs

```diff
@@ -2,33 +2,74 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+
 namespace System.Threading
 {
     /// <summary>
     /// A LIFO semaphore.
     /// Waits on this semaphore are uninterruptible.
     /// </summary>
-    internal sealed class LowLevelLifoSemaphore : IDisposable
+    internal sealed partial class LowLevelLifoSemaphore : IDisposable
     {
-        private WaitSubsystem.WaitableObject _semaphore;
+        private WaiterListEntry _waiterStackHead;
+        private LowLevelLock _waiterStackLock;
+        [ThreadStatic]
+        private static WaiterListEntry t_waitEntry;
 
-        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        private void Create(int maximumSignalCount)
         {
-            _semaphore = WaitSubsystem.WaitableObject.NewSemaphore(initialSignalCount, maximumSignalCount);
+            _waiterStackHead = null;
+            _waiterStackLock = new LowLevelLock();
         }
 
-        public bool Wait(int timeoutMs)
+        public void Dispose()
         {
-            return WaitSubsystem.Wait(_semaphore, timeoutMs, false, true);
         }
 
-        public int Release(int count)
+        private bool WaitForWake(int timeoutMs)
         {
-            return WaitSubsystem.ReleaseSemaphore(_semaphore, count);
+            WaiterListEntry waitEntry = t_waitEntry ?? (t_waitEntry = new WaiterListEntry());
```

Comment:
I wasn't happy with the implementation I took and I was considering to preallocate the sufficient number of entries. Unfortunately I currently don't know that number and would have to explicitly pass it through from the upper layers, which I wanted to avoid.

Reply:
The out-of-memory case can be prevented by having one "backup" `WaiterListEntry`. When a new one cannot be allocated you would just use that one and fallback to the FIFO implied order of the associated `LowLevelMonitor`. The logic is not trivial, but it's not too hard either.

Instead of pre-allocating all the entires they could simply be allocated on demand then and stored for reuse.

Reply:
For the thread pool, `maximumSignalCount` would be 32767 on 64-bit platforms, that's too many to preallocate. But the thread pool tracks how many max threads it intends to have at any given time, in the counts in `ThreadCounts.numThreadsGoal`. When the thread pool chooses to create threads you could check the goal value and ensure that at least that many have been allocated, if allocation fails it could be treated similarly to a thread creation failure.

One backup entry may not be enough, the thread pool may release more than one waiting thread at a time.

Reply:
> For the thread pool, `maximumSignalCount` would be 32767 on 64-bit platforms, that's too many to preallocate.

Agreed.

> When the thread pool chooses to create threads you could check the goal value and ensure that at least that many have been allocated, if allocation fails it could be treated similarly to a thread creation failure.

Possible. Violates the layering a bit, but not a deal-breaker.

> One backup entry may not be enough, the thread pool may release more than one waiting thread at a time.

I wasn't very clear about it. I meant that I only need single `LowLevelMonitor` to fallback to a FIFO behavior instead of LIFO. That can be easily pre-allocated when the semaphore is created.

(Another option I was considering is moving the `ReleaseCore`/`WaitCore` into unmanaged code and just allocating the list entry on stack.)

Reply:
This is what I meant with the OOM handling: https://gist.github.com/filipnavara/86cf7b27f0f20d86d6ebde9aebf158e1 (not tested)

Reply:
I see, that looks like it would work